cpp/src/arrow/compare.cc

```diff
@@ -1532,11 +1533,14 @@ bool DoubleEquals(const double& left, const double& right, const EqualOptions& o
   return result;
 }
 
-bool ArrayStatisticsValueTypeEquals(
-    const std::optional<ArrayStatistics::ValueType>& left,
-    const std::optional<ArrayStatistics::ValueType>& right, const EqualOptions& options) {
+template <typename Type>
+bool ArrayStatisticsValueTypeEquals(const std::optional<Type>& left,
```

Comment:
How about `ArrayStatisticsOptionalVariantEquals()`?

```suggestion
bool ArrayStatisticsOptionalVariantEquals(const std::optional<Type>& left,
```

Reply:
My suggestion is to change it to ArrayStatisticsValueEquals. Do you think this is better? I agree that my initial name was misleading

Reply:
Ah, https://github.com/apache/arrow/blob/f9301c0ba8a7ed1b0b63275cfdd4c44c26b04675/cpp/src/arrow/compare.cc#L1567-L1568 uses `std::optional<double>` not `std::optional<ValueTYpe>` by using `template`.

How about `ArrayStatisticsOptionalEquals()` or `ArrayStatisticsOptionalValueEquals()`? I think that `optional` is an important information for this.

Reply:
I think this one is better `ArrayStatisticsOptionalValueEquals()`